### PR TITLE
DAL focused on Google Calendar

### PR DESCRIPTION
**Pull request by madsmarcussen**
Not yet tested, but functionality is more or less complete.

### CODE REVIEW

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                string credPath = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
+                credPath = Path.Combine(credPath, ".credentials/calendar.json");
+
+                credential = GoogleWebAuthorizationBroker.AuthorizeAsync(
+                    GoogleClientSecrets.Load(stream).Secrets,
+                    Scopes,
+                    "user",
+                    CancellationToken.None,
+                    new FileDataStore(credPath, true)).Result;
+                //Console.WriteLine("Credential file saved to: " + credPath);
```
Comment (alminde7) [inline TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs]: Log til elasticseach

File: TaskSharper.DataAccessLayer.Google/Calendar/BaseService.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+using TaskSharper.Shared.Logging;
+
+namespace TaskSharper.DataAccessLayer.Google.Calendar
+{
+    public class BaseService
+    {
+        protected UserCredential UserCredential;
+        protected ILogger Logger;
+
+        public BaseService()
```
Comment (alminde7) [inline TaskSharper.DataAccessLayer.Google/Calendar/BaseService.cs]: Consider Dependency inject Logger and Google Authentication

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Calendar.v3;
+using Google.Apis.Util.Store;
+
+namespace TaskSharper.DataAccessLayer.Google.Authentication
+{
+    public static class GoogleAuthentication
```
Comment (alminde7) [inline TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs]: Consider making this class and methods not static for better testability

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs (hunk 1 above)
Comment (jeppekikt) [inline TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs]: Should this Console.writeline still be there?

File: TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+using Google.Apis.Calendar.v3;
+using Google.Apis.Services;
+using TaskSharper.Domain.Calendar;
+
+namespace TaskSharper.DataAccessLayer.Google.Calendar.Service
+{
+    public class GoogleCalendarService : BaseService, ICalendarService
+    {
+        private readonly CalendarService _service;
+        public GoogleCalendarService(CalendarService service = null)
```
Comment (alminde7) [inline TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs]: Why is CalendarService nullable.? Aren't we highly dependent on it. Consider making it not nullable, and therby forcing the user to create it.

File: TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        {
+            Authenticate();
+            if (service == null)
+            {
+                _service = new CalendarService(new BaseClientService.Initializer
+                {
+                    HttpClientInitializer = UserCredential,
+                    ApplicationName = "TaskSharper"
```
Comment (alminde7) [inline TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs]: Maybe consider to seperate the value "TaskSharper" into a _const string_ defined in the class, OR making a seperate .cs file (Constants.cs), defining it there and making the value static?

**Review by jeppekikt — APPROVED** [review]
Ser godt ud mads ;) !!

File: TaskSharper.Domain/Calendar/Event.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        /// </summary>
+        public EventStatus Status { get; set; }
+
+        /// <summary>
+        /// Creation time of the event. Read-only.
+        /// </summary>
+        public DateTime? Created { get; set; }
+
+        /// <summary>
+        /// First start time for a recurring event. Optional. TODO: Should this be included?
```
Comment (alminde7) [inline TaskSharper.Domain/Calendar/Event.cs]: Yes, let it be like this :)

**Review by alminde7 — CHANGES_REQUESTED** [review]
Looks good, there is minor architectural changes I think you should look into :)

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        private static readonly string[] Scopes = { CalendarService.Scope.Calendar };
+        public static UserCredential Authenticate()
+        {
+            UserCredential credential;
+
+            using (var stream = new FileStream("client_secret.json", FileMode.Open, FileAccess.Read))
+            {
+                string credPath = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
```
Comment (alminde7) [inline TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs:27]: This i fine for now, but at some point I think we have to consider an appropriate location for the client information to reside.

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs (hunk 1 above)
Comment (madsmarcussen) [reply TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs]: Done

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs (hunk 3 above)
Comment (madsmarcussen) [reply TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs]: Done

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs (hunk 1 above)
Comment (madsmarcussen) [reply TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs]: Done

File: TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs (hunk 5 above)
Comment (madsmarcussen) [reply TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs]: Made Constants.cs and inserted the value there.

File: TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs (hunk 7 above)
Comment (madsmarcussen) [reply TaskSharper.DataAccessLayer.Google/Authentication/GoogleAuthentication.cs:27]: Agreed.

File: TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs (hunk 4 above)
Comment (madsmarcussen) [reply TaskSharper.DataAccessLayer.Google/Calendar/Service/GoogleCalendarService.cs]: Didn't make sense for the user to create it. Changed the parameter to string applicationName.

**Review by alminde7 — APPROVED** [review]
Looking good

**Pull request merged**